File: Analysis/CorrelationAnalyzer.swift
import Foundation

/// Discovers cross-metric correlations and generates insights about how health behaviors connect
struct CorrelationAnalyzer {

    /// A pre-defined metric pair to test for correlation
    private struct MetricPair {
        let metricA: HealthMetric
        let metricB: HealthMetric
        let dayOffset: Int          // 0 = same day, 1 = A day N -> B day N+1
        let causeLabel: String      // "More deep sleep"
        let effectLabel: String     // "Higher HRV next day"
    }

    // MARK: - Comprehensive Pair Registry (~35 physiologically-plausible pairs)

    private static let pairs: [MetricPair] = [
        // Sleep -> Recovery (next-day offsets)
        MetricPair(metricA: .sleepDuration, metricB: .restingHeartRate, dayOffset: 1,
                   causeLabel: "More sleep", effectLabel: "Lower resting HR next day"),
        MetricPair(metricA: .sleepDuration, metricB: .heartRateVariability, dayOffset: 1,
                   causeLabel: "More sleep", effectLabel: "Higher HRV next day"),
        MetricPair(metricA: .sleepDeep, metricB: .heartRateVariability, dayOffset: 1,
                   causeLabel: "More deep sleep", effectLabel: "Higher HRV next day"),
        MetricPair(metricA: .sleepREM, metricB: .heartRateVariability, dayOffset: 1,
                   causeLabel: "More REM sleep", effectLabel: "Higher HRV next day"),
        MetricPair(metricA: .sleepDuration, metricB: .bloodOxygen, dayOffset: 1,
                   causeLabel: "More sleep", effectLabel: "Better blood oxygen next day"),

        // Exercise -> Recovery
        MetricPair(metricA: .exerciseMinutes, metricB: .heartRateVariability, dayOffset: 1,
                   causeLabel: "More exercise", effectLabel: "Higher HRV next day"),
        MetricPair(metricA: .activeCalories, metricB: .restingHeartRate, dayOffset: 1,
                   causeLabel: "More active calories", effectLabel: "Lower resting HR next day"),
        MetricPair(metricA: .activeCalories, metricB: .sleepDeep, dayOffset: 0,
                   causeLabel: "More active calories", effectLabel: "More deep sleep"),
        MetricPair(metricA: .exerciseMinutes, metricB: .sleepDuration, dayOffset: 0,
                   causeLabel: "More exercise", effectLabel: "Longer sleep"),
        MetricPair(metricA: .steps, metricB: .sleepDuration, dayOffset: 0,
                   causeLabel: "More steps", effectLabel: "Longer sleep"),
        MetricPair(metricA: .vo2Max, metricB: .restingHeartRate, dayOffset: 0,
                   causeLabel: "Higher VO2 Max", effectLabel: "Lower resting HR"),

        // Activity -> Body
        MetricPair(metricA: .steps, metricB: .activeCalories, dayOffset: 0,
                   causeLabel: "More steps", effectLabel: "More calories burned"),
        MetricPair(metricA: .standHours, metricB: .activeCalories, dayOffset: 0,
                   causeLabel: "More stand hours", effectLabel: "More calories burned"),
        MetricPair(metricA: .exerciseMinutes, metricB: .bodyFatPercentage, dayOffset: 0,
                   causeLabel: "More exercise", effectLabel: "Lower body fat"),
        MetricPair(metricA: .steps, metricB: .weight, dayOffset: 0,
                   causeLabel: "More steps", effectLabel: "Lower weight"),

        // Heart -> Stress / Mindfulness
        MetricPair(metricA: .mindfulMinutes, metricB: .heartRateVariability, dayOffset: 0,
                   causeLabel: "More mindfulness", effectLabel: "Higher HRV"),
        MetricPair(metricA: .mindfulMinutes, metricB: .restingHeartRate, dayOffset: 0,
                   causeLabel: "More mindfulness", effectLabel: "Lower resting HR"),
        MetricPair(metricA: .respiratoryRate, metricB: .heartRateVariability, dayOffset: 0,
                   causeLabel: "Lower respiratory rate", effectLabel: "Higher HRV"),

        // Sleep -> Activity (next-day)
        MetricPair(metricA: .sleepDuration, metricB: .steps, dayOffset: 1,
                   causeLabel: "More sleep", effectLabel: "More steps next day"),
        MetricPair(metricA: .sleepDuration, metricB: .exerciseMinutes, dayOffset: 1,
                   causeLabel: "More sleep", effectLabel: "More exercise next day"),
        MetricPair(metricA: .sleepDeep, metricB: .activeCalories, dayOffset: 1,
                   causeLabel: "More deep sleep", effectLabel: "More calories burned next day"),

        // Body -> Activity
        MetricPair(metricA: .weight, metricB: .restingHeartRate, dayOffset: 0,
                   causeLabel: "Higher weight", effectLabel: "Higher resting HR"),
        MetricPair(metricA: .bmi, metricB: .bloodPressureSystolic, dayOffset: 0,
                   causeLabel: "Higher BMI", effectLabel: "Higher blood pressure"),
        MetricPair(metricA: .bodyFatPercentage, metricB: .vo2Max, dayOffset: 0,
                   causeLabel: "Lower body fat", effectLabel: "Higher VO2 Max"),

        // Respiratory -> Sleep
        MetricPair(metricA: .respiratoryRate, metricB: .sleepDeep, dayOffset: 0,
                   causeLabel: "Lower respiratory rate", effectLabel: "More deep sleep"),
        MetricPair(metricA: .bloodOxygen, metricB: .sleepDeep, dayOffset: 0,
                   causeLabel: "Better blood oxygen", effectLabel: "More deep sleep"),

        // Additional cross-domain pairs
        MetricPair(metricA: .sleepDuration, metricB: .restingHeartRate, dayOffset: 0,
                   causeLabel: "More sleep", effectLabel: "Lower resting HR"),
        MetricPair(metricA: .sleepDeep, metricB: .restingHeartRate, dayOffset: 1,
                   causeLabel: "More deep sleep", effectLabel: "Lower resting HR next day"),
        MetricPair(metricA: .exerciseMinutes, metricB: .sleepDeep, dayOffset: 0,
                   causeLabel: "More exercise", effectLabel: "More deep sleep"),
        MetricPair(metricA: .steps, metricB: .heartRateVariability, dayOffset: 0,
                   causeLabel: "More steps", effectLabel: "Higher HRV"),
        MetricPair(metricA: .activeCalories, metricB: .heartRateVariability, dayOffset: 1,
                   causeLabel: "More active calories", effectLabel: "Higher HRV next day"),
        MetricPair(metricA: .sleepREM, metricB: .restingHeartRate, dayOffset: 1,
                   causeLabel: "More REM sleep", effectLabel: "Lower resting HR next day"),
        MetricPair(metricA: .vo2Max, metricB: .heartRateVariability, dayOffset: 0,
                   causeLabel: "Higher VO2 Max", effectLabel: "Higher HRV"),
        MetricPair(metricA: .exerciseMinutes, metricB: .restingHeartRate, dayOffset: 1,
                   causeLabel: "More exercise", effectLabel: "Lower resting HR next day"),
    ]

    // MARK: - Public API

    /// Analyze all metric pairs with automatic lag discovery (tests lags 0-3 for each pair)
    /// and return significant correlations as HealthCorrelation objects
    static func analyzeAll(timeSeries: [HealthMetric: MetricTimeSeries]) -> [HealthCorrelation] {
        var results: [HealthCorrelation] = []

        for pair in pairs {
            guard let seriesA = timeSeries[pair.metricA],
                  let seriesB = timeSeries[pair.metricB] else { continue }

            // Test the predefined lag AND nearby lags to find the strongest correlation
            let lagsToTest = Set([pair.dayOffset, 0, 1])

            var bestResult: (r: Double, t: Double, lag: Int, aligned: [TimeSeriesAligner.AlignedPair])?

            for lag in lagsToTest {
                let aligned: [TimeSeriesAligner.AlignedPair]
                if lag == 0 {
                    aligned = TimeSeriesAligner.alignByDate(seriesA, seriesB)
                } else {
                    aligned = TimeSeriesAligner.alignWithOffset(seriesA, seriesB, dayOffset: lag)
                }

                guard aligned.count >= 14 else { continue }

                guard let r = pearsonCorrelation(aligned),
                      abs(r) >= 0.2 else { continue }

                let n = Double(aligned.count)
                let rSquared = r * r
                guard rSquared < 1.0 else { continue }
                let t = abs(r) * sqrt((n - 2) / (1 - rSquared))
                guard t >= 2.0 else { continue }

                if bestResult == nil || abs(r) > abs(bestResult!.r) {
                    bestResult = (r: r, t: t, lag: lag, aligned: aligned)
                }
            }

            guard let best = bestResult else { continue }

            // Effect size: require >= 8% difference between above/below baseline groups
            let effectResult = computeConditionalEffect(
                aligned: best.aligned,
                metricA: pair.metricA,
                metricB: pair.metricB
            )
            guard effectResult.percentDiff >= 8.0 else { continue }

            // Update labels if the discovered lag differs from the predefined one
            let effectLabel: String
            if best.lag != pair.dayOffset && best.lag > 0 {
                let baseName = pair.effectLabel.replacingOccurrences(of: " next day", with: "")
                effectLabel = best.lag == 1 ? "\(baseName) next day" : "\(baseName) after \(best.lag) days"
            } else {
                effectLabel = pair.effectLabel
            }

            let strengthLabel = Self.strengthLabel(for: abs(best.r))

            // Build personalized labels using the user's actual data
            let baseline = best.aligned.mean(of: \.valueA)
            let personalCauseLabel = Self.personalizedCauseLabel(
                metric: pair.metricA,
                baseline: baseline
            )
            let personalEffectLabel = Self.personalizedEffectLabel(
                metric: pair.metricB,
                avgAbove: effectResult.avgBAbove,
                avgBelow: effectResult.avgBBelow,
                percentDiff: effectResult.percentDiff,
                dayOffset: best.lag
            )
            let personalSummary = Self.personalizedSummary(
                metricA: pair.metricA,
                metricB: pair.metricB,
                baseline: baseline,
                avgBAbove: effectResult.avgBAbove,
                avgBBelow: effectResult.avgBBelow,
                percentDiff: effectResult.percentDiff,
                dayOffset: best.lag,
                sampleCount: best.aligned.count,
                r: best.r
            )

            results.append(HealthCorrelation(
                metricA: pair.metricA,
                metricB: pair.metricB,
                correlation: best.r,
                sampleCount: best.aligned.count,
                strengthLabel: strengthLabel,
                causeLabel: personalCauseLabel,
                effectLabel: personalEffectLabel,
                effectSummary: personalSummary,
                effectPercentDiff: effectResult.percentDiff,
                isPositive: best.r > 0,
                dayOffset: best.lag,
                avgBAbove: effectResult.avgBAbove,
                avgBBelow: effectResult.avgBBelow
            ))
        }

        return results.sorted { abs($0.correlation) > abs($1.correlation) }
    }

    /// Generate top correlation insights with natural language text
    static func generateInsights(from correlations: [HealthCorrelation]) -> [Insight] {
        // Prioritize non-obvious correlations: cross-day > cross-category > same-category
        let ranked = correlations.sorted { a, b in
            let aScore = insightPriority(a)
            let bScore = insightPriority(b)
            return aScore > bScore
        }

        return Array(ranked.prefix(3)).map { result in
            let severity: Severity = abs(result.correlation) >= 0.5 ? .warning : .info

            return Insight(
                metric: result.metricA,
                title: insightTitle(result),
                summary: result.effectSummary,
                recommendation: insightRecommendation(result),
                severity: severity,
                trend: .stable,
                currentValue: result.correlation,
                baselineValue: 0,
                deviationPercent: abs(result.correlation) * 100,
                category: .correlation,
                relatedMetrics: [result.metricA, result.metricB]
            )
        }
    }

    /// Score how "non-obvious" and valuable a correlation is.
    /// Cross-day effects > cross-category > high effect size > same-category obvious ones.
    private static func insightPriority(_ c: HealthCorrelation) -> Double {
        var score = abs(c.correlation) * 10   // base: correlation strength
        score += c.effectPercentDiff * 0.3    // bigger effect = more interesting
        if c.dayOffset > 0 { score += 15 }    // lagged = non-obvious
        if c.metricA.category != c.metricB.category { score += 10 } // cross-category = surprising
        // Penalize obvious pairs
        let obviousPairs: Set<String> = ["steps-activeCalories", "activeCalories-steps", "standHours-activeCalories"]
        let pairKey = "\(c.metricA.rawValue)-\(c.metricB.rawValue)"
        if obviousPairs.contains(pairKey) { score -= 20 }
        return score
    }

    /// Natural language insight title. reads like a discovery, not a formula.
    private static func insightTitle(_ c: HealthCorrelation) -> String {
        let diffPct = Int(c.effectPercentDiff)
        let bFormatted = c.metricB.formatValue(c.avgBAbove)

        if c.dayOffset > 0 {
            // Lagged correlations. the most valuable ones
            return "Your \(c.metricB.displayName.lowercased()) is \(diffPct)% \(c.isPositive ? "higher" : "lower") the day after more \(c.metricA.displayName.lowercased())"
        }

        // Same-day correlations
        return "Days with more \(c.metricA.displayName.lowercased()) show \(diffPct)% \(c.isPositive ? "higher" : "lower") \(c.metricB.displayName.lowercased()) (\(bFormatted)\(c.metricB.unit))"
    }

    /// Conversational recommendation. what this means for the user.
    private static func insightRecommendation(_ c: HealthCorrelation) -> String {
        let aName = c.metricA.displayName.lowercased()
        let bName = c.metricB.displayName.lowercased()
        let formattedAbove = c.metricB.formatValue(c.avgBAbove)
        let formattedBelow = c.metricB.formatValue(c.avgBBelow)
        let diffPct = Int(c.effectPercentDiff)

        if c.dayOffset > 0 {
            return "In your data, higher \(aName) days are followed by \(bName) of \(formattedAbove)\(c.metricB.unit) vs \(formattedBelow)\(c.metricB.unit). a \(diffPct)% difference that shows up the next day. This pattern held across \(c.sampleCount) days."
        }

        return "On your above-average \(aName) days, \(bName) reaches \(formattedAbove)\(c.metricB.unit) instead of \(formattedBelow)\(c.metricB.unit). That \(diffPct)% gap is consistent across \(c.sampleCount) days of your data."
    }

    /// Legacy entry point. kept for backward compatibility
    static func generateInsights(timeSeries: [HealthMetric: MetricTimeSeries]) -> [Insight] {
        let correlations = analyzeAll(timeSeries: timeSeries)
        return generateInsights(from: correlations)
    }

    // MARK: - Private

    /// Strength label based on |r|
    private static func strengthLabel(for absR: Double) -> String {
        if absR >= 0.6 { return Copy.Analysis.Correlation.strong }
        if absR >= 0.4 { return Copy.Analysis.Correlation.moderate }
        return Copy.Analysis.Correlation.mild
    }

    /// Pearson correlation on aligned pairs without intermediate arrays.
    private static func pearsonCorrelation(_ aligned: [TimeSeriesAligner.AlignedPair]) -> Double? {
        let count = aligned.count
        guard count >= 5 else { return nil }

        var sumA = 0.0
        var sumB = 0.0
        for pair in aligned {
            sumA += pair.valueA
            sumB += pair.valueB
        }
        let meanA = sumA / Double(count)
        let meanB = sumB / Double(count)

        var numerator = 0.0
        var denomA = 0.0
        var denomB = 0.0
        for pair in aligned {
            let deltaA = pair.valueA - meanA
            let deltaB = pair.valueB - meanB
            numerator += deltaA * deltaB
            denomA += deltaA * deltaA
            denomB += deltaB * deltaB
        }

        guard denomA > 0, denomB > 0 else { return nil }
        return numerator / (denomA.squareRoot() * denomB.squareRoot())
    }

    // MARK: - Personalized Natural Language

    /// Cause label with the user's actual threshold number.
    /// "More exercise" → "40+ min of exercise"
    private static func personalizedCauseLabel(metric: HealthMetric, baseline: Double) -> String {
        let formatted = metric.formatValue(baseline)
        return "\(formatted)+ \(metric.displayName.lowercased())"
    }

    /// Effect label with the user's actual outcome numbers.
    /// "Higher HRV next day" → "HRV jumps to 45ms"
    private static func personalizedEffectLabel(
        metric: HealthMetric,
        avgAbove: Double,
        avgBelow: Double,
        percentDiff: Double,
        dayOffset: Int
    ) -> String {
        let formatted = metric.formatValue(avgAbove)
        let diffPct = Int(abs(percentDiff))
        let timing = dayOffset > 0 ? " next day" : ""
        return "\(metric.displayName) \(diffPct)% \(percentDiff > 0 ? "higher" : "lower")\(timing) (\(formatted)\(metric.unit))"
    }

    /// Natural language summary that reads like a coach, not a stats report.
    private static func personalizedSummary(
        metricA: HealthMetric,
        metricB: HealthMetric,
        baseline: Double,
        avgBAbove: Double,
        avgBBelow: Double,
        percentDiff: Double,
        dayOffset: Int,
        sampleCount: Int,
        r: Double
    ) -> String {
        let aName = metricA.displayName.lowercased()
        let bName = metricB.displayName.lowercased()
        let threshold = metricA.formatValue(baseline)
        let formattedAbove = metricB.formatValue(avgBAbove)
        let formattedBelow = metricB.formatValue(avgBBelow)
        let diffPct = Int(abs(percentDiff))
        let timing = dayOffset > 0 ? "the next morning" : "that same day"

        // Build a conversational explanation
        var parts: [String] = []

        // Opening. the finding in plain language
        parts.append("On days you hit \(threshold)+ \(aName), your \(bName) averages \(formattedAbove)\(metricB.unit) \(timing). compared to \(formattedBelow)\(metricB.unit) on lighter days.")

        // The "so what". why this matters
        parts.append("That's a \(diffPct)% difference.")

        // Confidence note. builds trust
        if sampleCount >= 30 {
            parts.append("This is based on \(sampleCount) days of your data. a reliable pattern.")
        } else {
            parts.append("Seen across \(sampleCount) days so far. still building confidence.")
        }

        return parts.joined(separator: " ")
    }

    struct EffectResult {
        let percentDiff: Double
        let summary: String
        let avgBAbove: Double
        let avgBBelow: Double
    }

    /// Partition by above/below baseline of metric A, compute average difference in metric B
    static func computeConditionalEffect(
        aligned: [TimeSeriesAligner.AlignedPair],
        metricA: HealthMetric,
        metricB: HealthMetric
    ) -> EffectResult {
        guard !aligned.isEmpty else {
            return EffectResult(
                percentDiff: 0,
                summary: "\(metricA.displayName) correlates with \(metricB.displayName).",
                avgBAbove: 0,
                avgBBelow: 0
            )
        }

        let baseline = aligned.mean(of: \.valueA)

        var aboveSum = 0.0
        var belowSum = 0.0
        var aboveCount = 0
        var belowCount = 0

        for pair in aligned {
            if pair.valueA >= baseline {
                aboveSum += pair.valueB
                aboveCount += 1
            } else {
                belowSum += pair.valueB
                belowCount += 1
            }
        }

        guard aboveCount > 0, belowCount > 0 else {
            return EffectResult(
                percentDiff: 0,
                summary: "\(metricA.displayName) correlates with \(metricB.displayName).",
                avgBAbove: 0,
                avgBBelow: 0
            )
        }

        let avgBAbove = aboveSum / Double(aboveCount)
        let avgBBelow = belowSum / Double(belowCount)

        guard avgBBelow != 0 else {
            return EffectResult(
                percentDiff: 0,
                summary: "\(metricA.displayName) correlates with \(metricB.displayName).",
                avgBAbove: avgBAbove,
                avgBBelow: 0
            )
        }

        let percentDiff = ((avgBAbove - avgBBelow) / abs(avgBBelow)) * 100
        let direction = percentDiff > 0 ? "higher" : "lower"
        let absPercent = String(format: "%.0f", abs(percentDiff))

        let formattedAbove = metricB.formatValue(avgBAbove)
        let formattedBelow = metricB.formatValue(avgBBelow)

        let summary = "When your \(metricA.displayName.lowercased()) is above average, your \(metricB.displayName.lowercased()) averages \(formattedAbove)\(metricB.unit) vs \(formattedBelow)\(metricB.unit) (\(absPercent)% \(direction))."

        return EffectResult(percentDiff: abs(percentDiff), summary: summary, avgBAbove: avgBAbove, avgBBelow: avgBBelow)
    }
}

// MARK: - InsightAnalyzer Conformance

extension CorrelationAnalyzer: InsightAnalyzer {
    static var analyzerID: String { "correlation" }
    static var insightCategory: InsightCategory { .correlation }

    static func generateInsights(context: AnalysisContext) -> [Insight] {
        generateInsights(from: context.correlations)
    }
}

File: android-app/app/src/main/java/com/lasohealth/android/core/analysis/analyzers/CorrelationAnalyzer.kt
package com.lasohealth.android.core.analysis.analyzers

import com.lasohealth.android.core.analysis.AnalysisContext
import com.lasohealth.android.core.analysis.AnalysisInsight
import com.lasohealth.android.core.analysis.DailySample
import com.lasohealth.android.core.analysis.InsightAnalyzer
import com.lasohealth.android.core.analysis.InsightDirective
import com.lasohealth.android.core.analysis.InsightPriority
import com.lasohealth.android.core.model.HealthMetric
import java.util.Calendar
import java.util.UUID
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Discovers cross-metric correlations from the user's daily time-series data
 * and generates human-readable insights about how health behaviours connect.
 *
 * Tests ~30 physiologically plausible metric pairs (not blind N^2), with automatic
 * lag discovery (0-1 day offsets) and effect-size validation.
 *
 * Ported from iOS CorrelationAnalyzer.swift.
 */
class CorrelationAnalyzer : InsightAnalyzer {

    override val name: String = "CorrelationAnalyzer"

    override suspend fun generateInsights(context: AnalysisContext): List<AnalysisInsight> {
        val correlations = analyzeAll(context.timeSeries)
        return generateInsightsFromCorrelations(correlations)
    }

    // ----- Pair Registry -----

    private data class MetricPair(
        val metricA: HealthMetric,
        val metricB: HealthMetric,
        val dayOffset: Int,
        val causeLabel: String,
        val effectLabel: String,
    )

    private val pairs = listOf(
        // Sleep -> Recovery (next-day)
        MetricPair(HealthMetric.SLEEP_DURATION, HealthMetric.RESTING_HEART_RATE, 1, "More sleep", "Lower resting HR next day"),
        MetricPair(HealthMetric.SLEEP_DURATION, HealthMetric.HEART_RATE_VARIABILITY, 1, "More sleep", "Higher HRV next day"),
        MetricPair(HealthMetric.SLEEP_DEEP, HealthMetric.HEART_RATE_VARIABILITY, 1, "More deep sleep", "Higher HRV next day"),
        MetricPair(HealthMetric.SLEEP_REM, HealthMetric.HEART_RATE_VARIABILITY, 1, "More REM sleep", "Higher HRV next day"),
        MetricPair(HealthMetric.SLEEP_DURATION, HealthMetric.BLOOD_OXYGEN, 1, "More sleep", "Better blood oxygen next day"),
        // Exercise -> Recovery
        MetricPair(HealthMetric.EXERCISE_MINUTES, HealthMetric.HEART_RATE_VARIABILITY, 1, "More exercise", "Higher HRV next day"),
        MetricPair(HealthMetric.ACTIVE_CALORIES, HealthMetric.RESTING_HEART_RATE, 1, "More active calories", "Lower resting HR next day"),
        MetricPair(HealthMetric.ACTIVE_CALORIES, HealthMetric.SLEEP_DEEP, 0, "More active calories", "More deep sleep"),
        MetricPair(HealthMetric.EXERCISE_MINUTES, HealthMetric.SLEEP_DURATION, 0, "More exercise", "Longer sleep"),
        MetricPair(HealthMetric.STEPS, HealthMetric.SLEEP_DURATION, 0, "More steps", "Longer sleep"),
        MetricPair(HealthMetric.VO2_MAX, HealthMetric.RESTING_HEART_RATE, 0, "Higher VO2 Max", "Lower resting HR"),
        // Activity -> Body
        MetricPair(HealthMetric.STEPS, HealthMetric.ACTIVE_CALORIES, 0, "More steps", "More calories burned"),
        MetricPair(HealthMetric.STAND_HOURS, HealthMetric.ACTIVE_CALORIES, 0, "More stand hours", "More calories burned"),
        // Heart / Stress / Mindfulness
        MetricPair(HealthMetric.MINDFUL_MINUTES, HealthMetric.HEART_RATE_VARIABILITY, 0, "More mindfulness", "Higher HRV"),
        MetricPair(HealthMetric.MINDFUL_MINUTES, HealthMetric.RESTING_HEART_RATE, 0, "More mindfulness", "Lower resting HR"),
        MetricPair(HealthMetric.RESPIRATORY_RATE, HealthMetric.HEART_RATE_VARIABILITY, 0, "Lower respiratory rate", "Higher HRV"),
        // Sleep -> Activity (next-day)
        MetricPair(HealthMetric.SLEEP_DURATION, HealthMetric.STEPS, 1, "More sleep", "More steps next day"),
        MetricPair(HealthMetric.SLEEP_DURATION, HealthMetric.EXERCISE_MINUTES, 1, "More sleep", "More exercise next day"),
        MetricPair(HealthMetric.SLEEP_DEEP, HealthMetric.ACTIVE_CALORIES, 1, "More deep sleep", "More calories burned next day"),
        // Respiratory -> Sleep
        MetricPair(HealthMetric.RESPIRATORY_RATE, HealthMetric.SLEEP_DEEP, 0, "Lower respiratory rate", "More deep sleep"),
        MetricPair(HealthMetric.BLOOD_OXYGEN, HealthMetric.SLEEP_DEEP, 0, "Better blood oxygen", "More deep sleep"),
        // Additional cross-domain
        MetricPair(HealthMetric.SLEEP_DURATION, HealthMetric.RESTING_HEART_RATE, 0, "More sleep", "Lower resting HR"),
        MetricPair(HealthMetric.EXERCISE_MINUTES, HealthMetric.SLEEP_DEEP, 0, "More exercise", "More deep sleep"),
        MetricPair(HealthMetric.STEPS, HealthMetric.HEART_RATE_VARIABILITY, 0, "More steps", "Higher HRV"),
        MetricPair(HealthMetric.ACTIVE_CALORIES, HealthMetric.HEART_RATE_VARIABILITY, 1, "More active calories", "Higher HRV next day"),
        MetricPair(HealthMetric.VO2_MAX, HealthMetric.HEART_RATE_VARIABILITY, 0, "Higher VO2 Max", "Higher HRV"),
        MetricPair(HealthMetric.EXERCISE_MINUTES, HealthMetric.RESTING_HEART_RATE, 1, "More exercise", "Lower resting HR next day"),
    )

    // ----- Analysis -----

    private data class CorrelationResult(
        val metricA: HealthMetric,
        val metricB: HealthMetric,
        val r: Double,
        val sampleCount: Int,
        val causeLabel: String,
        val effectLabel: String,
        val avgBAbove: Double,
        val avgBBelow: Double,
        val effectPercentDiff: Double,
        val dayOffset: Int,
    )

    private fun analyzeAll(timeSeries: Map<HealthMetric, List<DailySample>>): List<CorrelationResult> {
        val results = mutableListOf<CorrelationResult>()

        for (pair in pairs) {
            val samplesA = timeSeries[pair.metricA] ?: continue
            val samplesB = timeSeries[pair.metricB] ?: continue

            val lagsToTest = setOf(pair.dayOffset, 0, 1)
            var bestResult: Triple<Double, Int, List<Pair<Double, Double>>>? = null

            for (lag in lagsToTest) {
                val aligned = alignSamples(samplesA, samplesB, lag)
                if (aligned.size < 14) continue

                val r = pearsonCorrelation(aligned) ?: continue
                if (abs(r) < 0.2) continue

                // T-test significance
                val n = aligned.size.toDouble()
                val rSquared = r * r
                if (rSquared >= 1.0) continue
                val t = abs(r) * sqrt((n - 2) / (1 - rSquared))
                if (t < 2.0) continue

                if (bestResult == null || abs(r) > abs(bestResult.first)) {
                    bestResult = Triple(r, lag, aligned)
                }
            }

            val (bestR, bestLag, aligned) = bestResult ?: continue

            // Effect size: require >= 8% difference between above/below baseline groups
            val effect = computeConditionalEffect(aligned)
            if (effect.percentDiff < 8.0) continue

            results += CorrelationResult(
                metricA = pair.metricA,
                metricB = pair.metricB,
                r = bestR,
                sampleCount = aligned.size,
                causeLabel = pair.causeLabel,
                effectLabel = pair.effectLabel,
                avgBAbove = effect.avgBAbove,
                avgBBelow = effect.avgBBelow,
                effectPercentDiff = effect.percentDiff,
                dayOffset = bestLag,
            )
        }

        return results.sortedByDescending { abs(it.r) }
    }

    private fun generateInsightsFromCorrelations(correlations: List<CorrelationResult>): List<AnalysisInsight> {
        return correlations.take(3).map { result ->
            val diffPct = fmt0(result.effectPercentDiff)
            val direction = if (result.r > 0) "higher" else "lower"
            val strength = strengthLabel(abs(result.r))

            val detail = "When your ${result.metricA.displayName.lowercase()} is above average, " +
                "your ${result.metricB.displayName.lowercase()} averages ${fmt1(result.avgBAbove)} " +
                "vs ${fmt1(result.avgBBelow)} ${result.metricB.unit} ($diffPct% $direction). " +
                "Based on ${result.sampleCount} days of data ($strength correlation, r=${String.format("%.2f", result.r)})."

            AnalysisInsight(
                id = UUID.randomUUID().toString(),
                title = "${result.causeLabel} \u2192 ${result.effectLabel}",
                detail = detail,
                metric = result.metricA,
                priority = if (abs(result.r) >= 0.5) InsightPriority.MEDIUM else InsightPriority.INFORMATIONAL,
                directive = InsightDirective.INFORMATIONAL,
                confidence = minOf(result.sampleCount / 90.0, 1.0),
            )
        }
    }

    // ----- Alignment -----

    private fun alignSamples(
        samplesA: List<DailySample>,
        samplesB: List<DailySample>,
        dayOffset: Int,
    ): List<Pair<Double, Double>> {
        val mapA = buildDayMap(samplesA)
        val mapB = buildDayMap(samplesB)
        val aligned = mutableListOf<Pair<Double, Double>>()
        for ((day, valueA) in mapA) {
            val targetDay = day + dayOffset.toLong() * DAY_MS
            val valueB = mapB[targetDay] ?: continue
            aligned += valueA to valueB
        }
        return aligned
    }

    private fun buildDayMap(samples: List<DailySample>): Map<Long, Double> {
        val map = mutableMapOf<Long, Double>()
        for (s in samples) {
            map[startOfDay(s.date)] = s.value
        }
        return map
    }

    // ----- Statistics -----

    private fun pearsonCorrelation(aligned: List<Pair<Double, Double>>): Double? {
        val n = aligned.size
        if (n < 5) return null

        val meanA = aligned.map { it.first }.average()
        val meanB = aligned.map { it.second }.average()

        var numerator = 0.0; var denomA = 0.0; var denomB = 0.0
        for ((a, b) in aligned) {
            val dA = a - meanA; val dB = b - meanB
            numerator += dA * dB; denomA += dA * dA; denomB += dB * dB
        }
        if (denomA <= 0 || denomB <= 0) return null
        return numerator / (sqrt(denomA) * sqrt(denomB))
    }

    private data class EffectResult(val percentDiff: Double, val avgBAbove: Double, val avgBBelow: Double)

    private fun computeConditionalEffect(aligned: List<Pair<Double, Double>>): EffectResult {
        if (aligned.isEmpty()) return EffectResult(0.0, 0.0, 0.0)
        val baselineA = aligned.map { it.first }.average()

        var aboveSum = 0.0; var aboveCount = 0
        var belowSum = 0.0; var belowCount = 0

        for ((a, b) in aligned) {
            if (a >= baselineA) { aboveSum += b; aboveCount++ }
            else { belowSum += b; belowCount++ }
        }

        if (aboveCount == 0 || belowCount == 0) return EffectResult(0.0, 0.0, 0.0)
        val avgAbove = aboveSum / aboveCount
        val avgBelow = belowSum / belowCount
        if (avgBelow == 0.0) return EffectResult(0.0, avgAbove, 0.0)

        val pctDiff = abs((avgAbove - avgBelow) / abs(avgBelow)) * 100.0
        return EffectResult(pctDiff, avgAbove, avgBelow)
    }

    private fun strengthLabel(absR: Double): String = when {
        absR >= 0.6 -> "strong"
        absR >= 0.4 -> "moderate"
        else -> "mild"
    }

    // ----- Helpers -----

    private fun startOfDay(epochMs: Long): Long {
        val cal = Calendar.getInstance().apply { timeInMillis = epochMs }
        cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }

    private fun fmt0(v: Double): String = String.format("%.0f", v)
    private fun fmt1(v: Double): String = String.format("%.1f", v)

    companion object {
        private const val DAY_MS = 86_400_000L
    }
}
